lib/pbench/server/api/resources/users_api.py

```diff
@@ -217,16 +217,13 @@ def post(self):
                 500, message="INTERNAL ERROR",
             )
 
-        # check if the user is already logged in, in case the request has a an authorization header included
+        # Log if the request has an authorization header included and user is still trying to create a new login
         # We would still proceed to login and return a new auth token to the user
-        if user == self.auth.token_auth.current_user():
-            self.logger.warning("User already logged in and trying to re-login")
-            if self.auth.token_auth.get_auth()["token"] == auth_token:
-                # If a user attempts to re-login immediately within a millisecond,
-                # the new auth token will be same as the one present in the header
-                # Therefore we will not fulfill the re-login request
-                self.logger.info("Attempted immediate re-login")
-                abort(403, message="Retry login after some time")
+        if self.auth.token_auth.current_user():
```

Comment:
Can you explain more about how this situation could arise?

Doesn't this mean the `POST /login HTTP/1.1` request would have an `Authorization` header on it?  If so, why wouldn't we just take that header and use that token once the username and password are verified?

Reply:
Yes this situation would arise if `Authorization` header is present in the request. When you say use that token do you mean returning that same token? If so, yeah that's probably one way to handle it so we don't create another token if there already is valid token in header. However, we discussed earlier that we would just return the new token if the user ask for one assuming user knows what they are doing. Here is the link to the discussion https://github.com/distributed-system-analysis/pbench/pull/1937#discussion_r545920589

Reply:
Is it really the case that a user who knows what they are doing would ever re-login after logging in?  Wouldn't they have to somehow re-typing the URL and all that.  This seems like a case we really shouldn't be handling.

Reply:
Yeah this is just an edge case scenario, potential behavior from a user. 
There are 2 options: 
- We could return the same token 
- Just ignore the auth header (if present) and proceed with a request

Do you prefer the first one?

Reply:
My take has always been that we should ignore the unnecessary Authorization header and simply generate a new token. That led to confusion because, if they do this in relatively quick succession, they could get a duplicate token.

My main concern is for testing/debugging, where it's common to run server commands though curl or Postman, or other test scripts; and less common to run a full workflow including logout. I would like to be able to handle these cases cleanly; I've never worked with a system that didn't.

The edge case that I think is unlikely is a workflow that repeats login without a logout _and_ presents previous Authorization tokens on later login attempts. I suspect that if that happens it's probably fine to return the Authorization token rather than trying to generate a new one. This won't help with two "unauthorized" login attempts in succession, and that's what I would like to support.

I don't want to query and return a live pre-existing token for the user; because during testing we might (hopefully will) run concurrent functional tests, probably under the same user, and each needs to have their own authorization session; we don't want the first one that logs out to invalidate the others. And even if they log in at the same time, they need unique authorization tokens that can each be stored and retrieved. We definitely need to support multiple concurrent authorization sessions from the same user.

Reply:
At the risk of muddying things, I'd like to come at this from a different angle.  The duration of these tokens is expected/intended to be relatively short, which means that a long-session user is going to need either a way to "refresh" a token or to "re-login" without, in either case, having logged out first.  So, from the server's perspective, what is this request going to look like?  (And, on the client side, what is going to trigger the refresh/relogin operation?)

When a user logs out, we can simply mark the token as being "expired in need of refresh".  If there happens to be a second session using the same username, then having the first session logged out will be no different than having had the second session's token expire naturally -- in either case, the client will need to refresh the token, and it should be able to handle that implicitly.

With that as the groundwork, I don't think it matters whether we hand out unique tokens for every login by a given user.  At our whim, we can keep several tokens going, or we can hand back the same one (perhaps with an updated expiration) every time.

(What am I missing, here?)

Reply:
Oh, one additional thought:  a login request without an Authorization token would be a "new session", while a login request _with_ a token could be a "refresh"...if the distinction is important.  If the distinction is not important, then, I'm with Dave, we should just ignore the token if it shows up on a login request.

Reply:
That actually would be one way to "refresh" a session, although there's an intriguing article @aquibbaig discovered that instead recommends that login return a "role based" refresh token usable _only_ for refreshing a session, which when presented to a special refresh API will return a new authorization token. (Sounds complicated, still not 100% sure I'm at peace with this idea.)

In any case, this is all out of scope for our current user/authentication system which creates long-life tokens that can be used until (relatively far out) expiration or explicit deletion via logout.

Reply:
I believe that our tokens are [currently configured with a duration of 60 minutes](https://github.com/distributed-system-analysis/pbench/blob/e12844cadb6e8d7195f7db55943bba82e522e1f2/server/lib/config/pbench-server-default.cfg#L33) (which is longer than I expected, but too short for a serious session).

I haven't read Aquib's article yet (I started to look at it when he first pointed it out, but it looked long and I moved on to other things); nevertheless, I have to say that it's not immediately obvious how having a second token which allows you to renew the first is better than having just one....

Reply:
Yes @webbnh the value for the token expiration is really arbitrary right now. We haven't came up with a fixed and more reliable number yet but its definitely going to be lot longer than that. 

About the refresh token API, the article is interesting and after digging a little bit more I came across quite a few article that seem to be suggesting a similar approach but its indeed a little complicated to understand in one read. Anyway, this topic is out of the scope of this PR.

Reply:
Yeah, we need to decide how we're going to manage authorization sessions with the dashboard and other potential future clients (including our own functional test environment). We don't have to decide that right now.

Reply:
Forgive me, but I still don't understand this situation.  Can you explain in the comment in the code how this can occur more clearly, and why we need to log an informational message?

Reply:
Ah, I originally put this log message here when I was doing some functional testing on user authentications. I am not sure how useful this log statement would be when we have actual users. The log statement only gets invoked when the Authorization header is present in the request. If the user already has a valid token but not presented in the request, this log statement wont be invoked (so its a bit misleading). I can completely take out this message if we dont need it.